File: Autosaver.m
/*
Copyright (C) 2010-2012 Kristian Duske

This file is part of TrenchBroom.

TrenchBroom is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

TrenchBroom is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
*/

#import "Autosaver.h"
#import "MapDocument.h"
#import "MapWriter.h"

@interface Autosaver (private)

- (int)backupNoOf:(NSString *)theFilename;
- (void)saveBackup;

@end

@implementation Autosaver (private)

- (int)backupNoOf:(NSString *)theFilename {
    NSString* basename = [theFilename stringByDeletingPathExtension];
    NSUInteger lastSpaceIndex = [basename rangeOfString:@" " options:NSBackwardsSearch].location;
    if (lastSpaceIndex == NSNotFound)
        return -1;
    
    int backupNo = [[basename substringFromIndex:lastSpaceIndex + 1] intValue];
    if (backupNo == 0)
        return -1;
    
    return backupNo;
}

- (void)saveBackup {
    [timer release];
    
    if (lastAction != nil && [map isDocumentEdited] && (lastSave == nil || [lastAction compare:lastSave] == NSOrderedDescending)) {
        NSDate* now = [NSDate date];
        if ([now timeIntervalSinceDate:lastAction] < idleInterval) {
            timer = [[NSTimer scheduledTimerWithTimeInterval:idleInterval target:self selector:@selector(saveBackup) userInfo:nil repeats:NO] retain];
            return;
        }
        
        NSURL* fileUrl = [map fileURL];
        if (fileUrl != nil) {
            NSFileManager* fileManager = [NSFileManager defaultManager];
            
            NSString* mapPath = [fileUrl path];
            NSString* mapFilename = [mapPath lastPathComponent];
            NSString* mapBasename = [mapFilename stringByDeletingPathExtension];
            NSString* mapDirPath = [mapPath stringByDeletingLastPathComponent];
            NSString* autosavePath = [mapDirPath stringByAppendingPathComponent:@"autosave"];
            
            BOOL exists, isDir;
            exists = [fileManager fileExistsAtPath:autosavePath isDirectory:&isDir];
            if (exists && !isDir) {
                NSLog(@"Unable to create autosave direction at '%@' because a file exists at that location", autosavePath);
                return;
            }
            
            if (!exists && ![fileManager createDirectoryAtPath:autosavePath withIntermediateDirectories:NO attributes:nil error:NULL]) {
                NSLog(@"Unable to create autosave direction at '%@'", autosavePath);
                return;
            }
            
            NSArray* contents = [fileManager contentsOfDirectoryAtPath:autosavePath error:NULL];
            if (contents == nil) {
                NSLog(@"Unable to access autosave direction at '%@'", autosavePath);
                return;
            }
            
            NSMutableArray* backups = [[NSMutableArray alloc] init];
            int highestBackupNo = 0;
            for (NSString* content in contents) {
                if ([content hasPrefix:mapBasename]) {
                    int backupNo = [self backupNoOf:content];
                    if (backupNo != -1) {
                        highestBackupNo = maxi(highestBackupNo, backupNo);
                        [backups addObject:content];
                    }
                }
            }
            
            int backupNo = 1;
            if ([backups count] > 0) {
                [backups sortUsingComparator:^(NSString* filename1, NSString* filename2) {
                    return [filename1 compare:filename2 options:NSNumericSearch];
                }];
                
                BOOL moveFiles = NO;
                for (int i = 0; i <= ((int)[backups count]) - numberOfBackups; i++) {
                    NSString* backupPath = [autosavePath stringByAppendingPathComponent:[backups objectAtIndex:0]];
                    if (![fileManager removeItemAtPath:backupPath error:NULL]) {
                        NSLog(@"could not delete autosave file '%@'", backupPath);
                        return;
                    }
                    
                    [backups removeObjectAtIndex:0];
                    moveFiles = YES;
                }
                
                if (moveFiles) {
                    for (int i = 0; i < [backups count]; i++) {
                        NSString* backupPath = [autosavePath stringByAppendingPathComponent:[backups objectAtIndex:i]];
                        NSString* newPath = [[autosavePath stringByAppendingPathComponent:[NSString stringWithFormat:@"%@ %i", mapBasename, i + 1]] stringByAppendingPathExtension:@"map"];
                        
                        if (![fileManager moveItemAtPath:backupPath toPath:newPath error:NULL]) {
                            NSLog(@"could not move autosave file '%@' to '%@'", backupPath, newPath);
                            return;
                        }
                    }
                }
                
                backupNo =  [backups count] + 1;
            }
            
            [backups release];
            
            NSString* backupBasename = [NSString stringWithFormat:@"%@ %i", mapBasename, backupNo];
            NSString* backupPath = [[autosavePath stringByAppendingPathComponent:backupBasename] stringByAppendingPathExtension:@"map"];
            
            NSLog(@"Autosaving to file '%@'", backupPath);
            
            MapWriter* mapWriter = [[MapWriter alloc] initWithMap:map];
            [mapWriter writeToFileAtPath:backupPath];
            [mapWriter release];
            
            lastSave = [[NSDate date] retain];
        }
    }
    
    timer = [[NSTimer scheduledTimerWithTimeInterval:saveInterval target:self selector:@selector(saveBackup) userInfo:nil repeats:NO] retain];
}

@end

@implementation Autosaver

- (id)initWithMap:(MapDocument *)theMap saveInterval:(NSTimeInterval)theSaveInterval idleInterval:(NSTimeInterval)theIdleInterval numberOfBackups:(int)theNumberOfBackups {
    NSAssert(theMap != nil, @"map must not be nil");
    NSAssert(theSaveInterval > 0, @"save interval must be greater than 0");
    NSAssert(theIdleInterval >= 0, @"idle interval must be greater than or equal to 0");
    NSAssert(theNumberOfBackups > 0, @"number of backups must be greater than 0");
    
    if ((self = [self init])) {
        map = theMap;
        saveInterval = theSaveInterval;
        idleInterval = theIdleInterval;
        numberOfBackups = theNumberOfBackups;
        timer = [[NSTimer scheduledTimerWithTimeInterval:saveInterval target:self selector:@selector(saveBackup) userInfo:nil repeats:NO] retain];
    }
    
    return self;
}

- (void)dealloc {
    [self terminate];
    [lastSave release];
    [lastAction release];
    [super dealloc];
}

- (void)updateLastAction {
    [lastAction release];
    lastAction = [[NSDate date] retain];
}

- (void)terminate {
    if (timer != nil) {
        [timer invalidate];
        [timer release];
        timer = nil;
    }
}

@end

File: Core/Source/Controller/Autosaver.cpp
/*
 Copyright (C) 2010-2012 Kristian Duske
 
 This file is part of TrenchBroom.
 
 TrenchBroom is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 TrenchBroom is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "Autosaver.h"

#include "Controller/Editor.h"
#include "IO/FileManager.h"
#include "Utilities/Console.h"

#include <string>
#include <cassert>
#include <cstdlib>
#include <algorithm>

namespace TrenchBroom {
    namespace Controller {
        unsigned int backupNoOfFile(const std::string& path) {
            IO::FileManager& fileManager = *IO::FileManager::sharedFileManager;
            std::string basePath = fileManager.deleteExtension(path);
            size_t spaceIndex = basePath.find_last_of(' ');
            if (spaceIndex == std::string::npos)
                return 0;
            
            int backupNo = std::atoi(basePath.substr(spaceIndex + 1).c_str());
            backupNo = std::max(backupNo, 0);
            return static_cast<unsigned int>(backupNo);
        }

        bool compareByBackupNo(const std::string& file1, const std::string& file2) {
            unsigned int backupNo1 = backupNoOfFile(file1);
            unsigned int backupNo2 = backupNoOfFile(file2);
            return backupNo1 < backupNo2;
        }

        void Autosaver::triggerAutosave() {
            time_t currentTime = time(NULL);
            if (m_dirty && m_lastModificationTime > 0 && currentTime - m_lastModificationTime >= m_idleInterval && currentTime - m_lastSaveTime >= m_saveInterval) {
                
                const std::string& mapPath = m_editor.mapPath();
                if (mapPath.empty())
                    return;
                
                IO::FileManager& fileManager = *IO::FileManager::sharedFileManager;
                std::string basePath = fileManager.deleteLastPathComponent(mapPath);
                std::string autosavePath = fileManager.appendPath(basePath, "autosave");
                std::string mapFilename = fileManager.pathComponents(mapPath).back();
                std::string mapBasename = fileManager.deleteExtension(mapFilename);
                
                if (!fileManager.exists(autosavePath)) {
                    if (!fileManager.makeDirectory(autosavePath)) {
                        log(TB_LL_ERR, "Cannot create autosave directory at %s", autosavePath.c_str());
                        return;
                    }
                    
                    log(TB_LL_INFO, "Autosave directory created at %s", autosavePath.c_str());
                } else if (!fileManager.isDirectory(autosavePath)) {
                    log(TB_LL_ERR, "Cannot create autosave directory at %s because a file exists at that path", autosavePath.c_str());
                    return;
                }
                
                // collect the actual backup files and determine the highest backup no
                std::vector<std::string> contents = fileManager.directoryContents(autosavePath);
                std::vector<std::string> backups;
                unsigned int highestBackupNo = 1;
                for (unsigned int i = 0; i < contents.size(); i++) {
                    const std::string& filename = contents[i];
                    if (filename.substr(0, mapBasename.length()) == mapBasename) {
                        unsigned int backupNo = backupNoOfFile(filename);
                        if (backupNo > 0) {
                            highestBackupNo = std::max(highestBackupNo, backupNo);
                            backups.push_back(filename);
                        }
                    }
                }
                
                if (!backups.empty()) {
                    // sort the backups by their backup nos
                    std::sort(backups.begin(), backups.end(), compareByBackupNo);
                    
                    if (highestBackupNo > backups.size()) {
                        // reorganize the backups and remove "holes"
                        // adapt highestBackupNo
                    }

                    assert(highestBackupNo == backups.size());
                    
                    if (backups.size() >= m_maxBackups) {
                        // remove the first backups until backups.size() == m_maxBackups - 1
                        // adapt highestBackupNo
                    }
                }
                
                // append the backup with file name "mapBasename highestBackupNo.map"
                
                
                m_lastSaveTime = currentTime;
                m_dirty = false;
            }
        }
        
        void Autosaver::updateLastModificationTime() {
            m_lastModificationTime = time(NULL);
            m_dirty = true;
        }

        void Autosaver::clearDirtyFlag() {
            m_dirty = false;
        }
    }
}